### PR TITLE
SNOW-1516310: [Local Testing] Add support for datediff

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1516310

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   This PR adds support for datediff in local test as well as adds a test that verifies more edge cases than the tests that were present.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -41,6 +41,7 @@
 - Added support for the following APIs:
   - snowflake.snowpark.functions
     - random
+    - datediff
```
Comment (sfc-gh-aling) [inline CHANGELOG.md:44]: nit: this goes to version 1.21

File: src/snowflake/snowpark/mock/_functions.py [hunk 2]
```diff
@@ -1694,6 +1694,44 @@ def mock_date_trunc(part: str, datetime_expr: ColumnEmulator) -> ColumnEmulator:
     return ColumnEmulator(truncated, sf_type=datetime_expr.sf_type)
 
 
+@patch("datediff")
+def mock_datediff(
+    part: str, col1: ColumnEmulator, col2: ColumnEmulator
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_functions.py:1699]: are col1 and col2 guaranteed to be of type timestamp -- is it possible that col1 is of string type while col2 is int type?

**Comment by sfc-gh-aling** [comment]
it looks like the live test is failing, can you check?

File: src/snowflake/snowpark/mock/_functions.py (hunk 2 above)
Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/mock/_functions.py:1699]: In sql both columns have to be either dates or timestamps. I could add a validation for column type, but I think we've avoided doing that for other functions so far.

**Review by sfc-gh-aling — APPROVED** [review]
